core/src/main/scala/besom/internal/codecs.scala

```diff
@@ -1027,91 +1124,110 @@ object ArgsEncoder:
+                case ( /* acc */ (mapOfMetadata, mapOfValues), /* fields */ (label, (metadata, value))) =>
+                  if filterOut(label) then (mapOfMetadata, mapOfValues)
+                  else if metadata.empty then
+                    (mapOfMetadata, mapOfValues) // we treat properties with NullValue values as if they do not exist
+                  else (mapOfMetadata + (label -> metadata), mapOfValues + (label -> value))
               }
 
-            val struct = mapOfValues.asStruct
-
-            mapOfResources -> struct
+            if (mapOfMetadata.keySet != mapOfValues.keySet)
+              Result.fail(
+                Exception(
+                  s"Post-condition: expected ProviderArgsEncoder to return metadata and values for the same keys, " +
+                    s"got: ${mapOfMetadata.keys.mkString(",")}, ${mapOfValues.keys.mkString(",")}"
+                )
+              )
+            else
+              Result.pure(mapOfMetadata -> mapOfValues.asStruct)
           }
 
-  inline def derived[A <: Product](using m: Mirror.ProductOf[A]): ProviderArgsEncoder[A] =
-    lazy val labels: List[String]            = CodecMacros.summonLabels[m.MirroredElemLabels]
-    lazy val instances: List[JsonEncoder[?]] = CodecMacros.summonJsonEncoders[m.MirroredElemTypes]
-    lazy val nameEncoderPairs                = labels.zip(instances)
-
-    providerArgsEncoderProduct(nameEncoderPairs)
-end ProviderArgsEncoder
-
-trait JsonEncoder[A]:
-  def encode(a: A): Result[(Set[Resource], Value)]
-
-object JsonEncoder:
-  import ProtobufUtil.*
-
-  given jsonEncoder[A](using enc: Encoder[A]): JsonEncoder[A] =
-    new JsonEncoder[A]:
-      def encode(a: A): Result[(Set[Resource], Value)] = enc.encode(a).flatMap {
-        case (resources, v @ Value(Kind.NullValue(_), _))   => Result.pure(resources -> v)
-        case (resources, s @ Value(Kind.StringValue(_), _)) => Result.pure(resources -> s)
-        // if the value is an empty secret, we don't want to serialize it as a JSON string, ProviderArgsEncoder will filter it out
-        case (resources, v @ Value(Kind.StructValue(_), _)) if Encoder.isEmptySecretValue(v) => Result.pure(resources -> v)
-        case (resources, value) =>
-          Result.evalEither(value.asJsonString).transform {
-            case Left(ex) =>
-              Left(Exception("Encountered a malformed protobuf Value that could not be serialized to JSON", ex))
-            case Right(jsonString) => Right(resources -> jsonString.asValue)
+  private def removeSpecialStructures(value: Value): Value =
+    value match
+      case OutputValue(o) => removeSpecialStructures(o.asRawValueOrUnknown)
+      case SecretValue(s) => removeSpecialStructures(s.value)
+      case Value(Kind.StructValue(s), _) =>
+        Struct(fields = s.fields.map {
+          case (k, v) => {
+            k -> removeSpecialStructures(v)
           }
-      }
+        }).asValue
+      case v => v
```

Comment:
missing Array of Struct here

Reply:
oh wow, good that you've caught this, thank you!